File: cpp/src/lib/download.cpp
#include "download.h"
#include "protocol/structs.h"
#include "util.h"

#include <algorithm>
#include <cassert>
#include <expected>
#include <fstream>
#include <iostream>
#include <unistd.h>
#include <vector>
// if destdir is non-null, the file contents will be read and written to
// disk
std::vector<FileInfo> read_file_list(FileDescriptor &fd, size_t file_count,
                                     const fs::path *destdir) {
  std::vector<FileInfo> list;
  { // recv the file info
    std::byte buf[4096];
    for (size_t i = 0; i < file_count; i++) {
      auto msg = read_message(fd, buf);
      protocol::download_file_m f;
      if (!protocol::deserialize(f, msg))
        throw std::runtime_error("error deserializing download file");
      list.emplace_back(std::string_view{f.name, f.name_len}, f.hash, f.size);
    }
  }
  assert(file_count == list.size());
  if (destdir) {
    // recv/write the file contents
    for (const FileInfo &iter : list) {
      const fs::path path = *destdir / iter.path; // resolve path according to destdir
      if (!fs::create_directories(path.parent_path()))
        throw std::runtime_error("error creating directory");
      std::ofstream ofs{path, std::ios::binary};
      if (!ofs) throw std::runtime_error("error opening file for writing\n");
      std::cout << "receiving " << path << ": " << iter.hash << std::endl;
      fd.read_to(ofs, iter.size);
      // TODO: verify hash (?)
    }
  }
  return list;
}

// if destdir is NULL, the files will not be read from the message (ie. the
// uploader must not send the file contents)
std::vector<FileInfo> read_download_message(FileDescriptor &fd, const fs::path *destdir) {
  std::byte buf[4096]; // recv download message
  const auto recv = read_message(fd, buf);
  protocol::download_m msg{};
  if (!deserialize(msg, recv)) throw std::runtime_error("error deserializing download message");
  return read_file_list(fd, msg.file_count, destdir);
}

void download(FileDescriptor &sockfd, std::span<FileInfo> files, const fs::path &destdir) {
  //  read the download message
  auto recvlist = read_download_message(sockfd, nullptr);

  const auto to_delete =
      files | std::views::filter([&recvlist](const FileInfo &f) {
        return std::ranges::none_of(recvlist, [&f](const FileInfo &o) { return o.path == f.path; });
      }) |
      std::ranges::to<std::vector<FileInfo>>();
  // filter the recv list to exclude anything that we already have
  std::erase_if(recvlist, [&files](const FileInfo &f) {
    return std::ranges::any_of(
        files, [&f](const FileInfo &o) { return o.path == f.path && o.hash == f.hash; });
  });

  {                                 // send download response
    assert(recvlist.size() <= 255); // this is a protocol limit
    const protocol::download_response_m resp{recvlist};
    std::byte buf[4096];
    auto end = serialize(buf, resp).value();
    const size_t len = std::distance(std::begin(buf), end.data());
    write_message(sockfd, std::span{buf, len});
  }

  // read download message 2
  read_download_message(sockfd, &destdir);

  // delete files that we don't need anymore
  for (const FileInfo &iter : to_delete) {
    fs::path p = destdir / iter.path;
    std::cout << "deleting " << p << std::endl;
    // remove the file and parent directories
    while (!p.empty()) {
      try {
        remove(p);
      } catch (const std::filesystem::filesystem_error &e) {
        if (e.code() == std::errc::directory_not_empty) break;
        throw;
      }
      p = p.parent_path();
    }
  }
}

File: rust/lib/src/download.rs
use std::path::Path;

use tokio::fs;
use tokio::io::{self, AsyncRead, AsyncReadExt, AsyncWrite};

use crate::download_file::DownloadFile;
use crate::download_message::{DownloadMessage, MessageDeserializeError};
use crate::download_response::{self, DownloadResponse};
use crate::file_hash::FileHash;
use crate::file_info::FileInfo;
use crate::protocol::{ProtocolError, read_message, write_message};
use crate::serial::{Deserialize, Serialize};

// if destdir is non-null, the file contents will be read and written to
// disk
async fn read_file_list(
    socket: &mut (dyn AsyncRead + Unpin + Send),
    file_count: u8,
    destdir: Option<&Path>,
) -> Result<Vec<FileInfo>, std::io::Error> {
    // recv the file info
    let mut list = Vec::<FileInfo>::with_capacity(file_count as usize);
    for _ in 0..file_count {
        let msg = read_message(socket).await?;
        let file = DownloadFile::deserialize(&mut &msg[..])?;
        list.push(file.into());
    }
    let list = list;

    // recv/write the file contents
    if let Some(destdir) = destdir {
        for f in &list {
            let path = destdir.join(f.path());
            fs::create_dir_all(path.parent().unwrap()).await?;
            let mut file = fs::File::create(&path).await?;
            io::copy(&mut socket.take(f.size()), &mut file).await?;
            // TODO: verify hash
        }
    }

    Ok(list)
}

#[derive(Debug, thiserror::Error)]
pub enum DownloadError {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error("deserialization error: {0}")]
    Deserialization(MessageDeserializeError),
    #[error("protocol error: {0}")]
    Protocol(ProtocolError),
}
impl From<MessageDeserializeError> for DownloadError {
    fn from(e: MessageDeserializeError) -> Self {
        match e {
            MessageDeserializeError::Io(e) => Self::Io(e),
            _ => Self::Deserialization(e),
        }
    }
}
impl From<ProtocolError> for DownloadError {
    fn from(e: ProtocolError) -> Self {
        match e {
            ProtocolError::Io(e) => Self::Io(e),
            _ => Self::Protocol(e),
        }
    }
}

// if destdir is NULL, the files will not be read from the message (ie. the
// uploader must not send the file contents)
async fn read_download_message(
    socket: &mut (dyn AsyncRead + Unpin + Send),
    destdir: Option<&Path>,
) -> Result<Vec<FileInfo>, DownloadError> {
    let msg = read_message(socket).await?;
    let msg = DownloadMessage::deserialize(&mut &msg[..])?;
    return Ok(read_file_list(socket, msg.file_count(), destdir).await?);
}

pub async fn download(
    read: &mut (dyn AsyncRead + Unpin + Send),
    write: &mut (dyn AsyncWrite + Unpin + Send),
    files: &Vec<FileInfo>,
    srcdir: &Path,
) -> Result<(), DownloadError> {
    //  read the download message
    let mut recvlist = read_download_message(read, None).await?;
    let to_delete = files
        .iter()
        .filter(|f| recvlist.iter().find(|o| o.path() == f.path()).is_none())
        .collect::<Vec<_>>();
    // filter the recv list to exclude anything that we already have; keep only the files that we need
    recvlist
        .retain(|f| files.iter().find(|o| o.path() == f.path() && o.hash() == f.hash()).is_none());
    {
        // send download response
        assert!(recvlist.len() <= 255); // this is a protocol limit
        let hashes: Vec<FileHash> = recvlist.iter().map(|f| f.hash()).cloned().collect();
        let resp = DownloadResponse::new(download_response::Flags::empty(), hashes);
        let mut buf = Vec::with_capacity(4096);
        resp.serialize(&mut buf)?;
        write_message(write, &buf).await?;
    }
    // read download message 2
    read_download_message(read, Some(srcdir)).await?;
    // delete files that we don't need anymore
    for node in to_delete {
        let mut path = srcdir.join(node.path());
        println!("deleting {}", path.display());
        // remove the file and parent directories
        fs::remove_file(&path).await?;
        while path.pop() && path != srcdir && !path.is_empty() {
            match fs::remove_dir(&path).await {
                Ok(_) => break,
                Err(e) if e.kind() == io::ErrorKind::NotFound => break,
                Err(e) if e.kind() == io::ErrorKind::DirectoryNotEmpty => break,
                Err(e) => return Err(e.into()),
            }
        }
    }
    Ok(())
}
